logstash-core/lib/logstash/agent.rb

```diff
@@ -11,32 +12,75 @@
 LogStash::Environment.load_locale!
 
 class LogStash::Agent
+  attr_reader :logger, :state
 
-  attr_writer :logger
-  attr_reader :node_name
-
-  def initialize(options = {})
+  def initialize(params)
+    @logger = params[:logger]
     @pipelines = {}
      
-    @node_name = options[:node_name] || Socket.gethostname
+    @node_name = params[:node_name] || Socket.gethostname
+    @pipeline_threads = {}
+    @config_loader = LogStash::Config::Loader.new(@logger, false)
+    @config_string = params[:config_string]
+    @config_path = params[:config_path]
+    @auto_reload = params[:auto_reload]
+    @pipeline_settings = (params[:pipeline_settings] || {})
+    @reload_interval = params[:reload_interval] || 3 # seconds
+    @upgrade_mutex = Mutex.new
+    # @state represents the necessary information for the agent's lifecycle.
+    # this base implementation uses the configuration string itself as state.
+    # if fetch_state returns a different string, it is regarded as new state
+    # and upgrade_state will take care of stopping the previous pipeline and
+    # starting a new one
+    # a subclass of this base Agent could use more complex state objects such
+    # as a hash map containing versioning, update_at timestamp, etc.
+    @state = clean_state
   end
 
   def execute
-    # Make SIGINT/SIGTERM shutdown the pipeline.
-    sigint_id = trap_sigint()
-    sigterm_id = trap_sigterm()
+    @thread = Thread.current # this var is implicilty used by Stud.stop?
+    @logger.info("starting agent", :state => @state)
+
+    reload_state!
+    return 1 if clean_state?
+
+    if @auto_reload
+      Stud.interval(@reload_interval) { reload_state! }
+    else
+      while !Stud.stop?
+        if clean_state? || running_pipelines?
+          sleep 0.5
+        else
+          break
+        end
+      end
+    end
+  end
+
+  def shutdown
+    shutdown_pipelines
+  end
 
-    @pipelines.each {|_, p| p.run } # blocking operation. works now because size <= 1
-    return 0
+  def reload_state!
+    new_state = fetch_state
+
+    if (reason = valid_state?(new_state)) != true
+      @logger.error("invalid state", :state => new_state, :reason => reason)
```

Comment:
Hmmm, will this always print in the `return 1` case? If so this is a fine error.

The confusing thing about using the 'state' verbiage, is that the user is dealing with config files. `state` is really an internal term. Thoughts?